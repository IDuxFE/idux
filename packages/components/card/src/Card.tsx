/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { CardProps } from './types'
import type { ComputedRef, Slots, VNode, VNodeTypes } from 'vue'

import { computed, defineComponent, isVNode, normalizeClass, provide } from 'vue'

import { isString } from 'lodash-es'

import { ɵHeader } from '@idux/components/_private/header'
import { IxButton } from '@idux/components/button'
import { useGlobalConfig } from '@idux/components/config'
import { IxCol, IxRow } from '@idux/components/grid'

import { cardToken } from './token'
import { cardProps } from './types'

export default defineComponent({
  name: 'IxCard',
  props: cardProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-card`)
    const config = useGlobalConfig('card')
    const hoverable = computed(() => props.hoverable ?? config.hoverable)

    provide(cardToken, { hoverable })

    const children = computed(() => slots.default?.() || [])
    const hasGrid = computed(() => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return children.value.some(node => node.type && (node.type as any).name === 'IxCardGrid')
    })

    const classes = computed(() => {
      const { borderless = config.borderless, loading, size = config.size } = props
      const hasGridValue = hasGrid.value
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-borderless`]: borderless,
        [`${prefixCls}-hoverable`]: !hasGridValue && hoverable.value,
        [`${prefixCls}-loading`]: loading,
        [`${prefixCls}-has-grid`]: hasGridValue,
        [`${prefixCls}-${size}`]: true,
      })
    })

    return () => {
      const prefixCls = mergedPrefixCls.value
      return (
        <div class={classes.value}>
          {renderCover(props, slots, prefixCls)}
          <ɵHeader v-slots={slots} header={props.header} />
          {renderBody(props, children, hasGrid, prefixCls)}
          {renderFooter(props, slots, prefixCls)}
        </div>
      )
    }
  },
})

const renderCover = (props: CardProps, slots: Slots, prefixCls: string) => {
  let coverNode: VNodeTypes | undefined
  if (slots.cover) {
    coverNode = slots.cover()
  } else if (props.cover) {
    const { cover } = props
    const imgProps = isString(cover) ? { src: cover } : cover
    coverNode = <img {...imgProps} />
  }
  return coverNode ? <div class={`${prefixCls}-cover`}>{coverNode}</div> : undefined
}

const renderBody = (
  props: CardProps,
  children: ComputedRef<VNode[]>,
  hasGrid: ComputedRef<boolean>,
  prefixCls: string,
) => {
  let bodyNode: VNodeTypes | undefined
  if (props.loading) {
    bodyNode = renderLoading(prefixCls)
  } else if (children.value.length) {
    bodyNode = hasGrid.value ? <IxRow>{children.value}</IxRow> : children.value
  }
  return bodyNode ? <div class={`${prefixCls}-body`}>{bodyNode}</div> : undefined
}

const loadingSpans = [[22], [8, 15], [6, 18], [13, 9], [4, 3, 16], [8, 6, 8]]

const renderLoading = (prefixCls: string) => {
  const loadingChild = loadingSpans.map(spans => {
    const cols = spans.map(span => (
      <IxCol span={span} class={`${prefixCls}-loading-col`}>
        <div class={`${prefixCls}-loading-block`}></div>
      </IxCol>
    ))
    return <IxRow>{cols}</IxRow>
  })
  return <div class={`${prefixCls}-loading`}>{loadingChild}</div>
}

const renderFooter = (props: CardProps, slots: Slots, prefixCls: string) => {
  if (!slots.footer && !props.footer) {
    return undefined
  }

  let footerNode: VNodeTypes
  if (slots.footer) {
    footerNode = slots.footer()
  } else {
    footerNode = props.footer!.map(item => {
      let itemChild: VNodeTypes | null
      if (isVNode(item)) {
        itemChild = item
      } else {
        const { text, ...rest } = item
        itemChild = <IxButton {...rest}>{text}</IxButton>
      }
      return itemChild ? <li>{itemChild}</li> : null
    })
  }

  return <ul class={`${prefixCls}-footer`}>{footerNode}</ul>
}
