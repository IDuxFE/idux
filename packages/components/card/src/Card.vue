<template>
  <div class="ix-card" :class="classes">
    <div v-if="isShowLoading" class="ix-card-loading-block"></div>
    <div v-else class="ix-card-wraper">
      <div class="ix-card__head">
        <div v-if="isShowTitle" class="ix-card__head__title">
          <slot name="title">
            <a>{{ title }}</a>
          </slot>
        </div>
        <div v-if="isShowExtra" class="ix-card__head__extra">
          <slot name="extra">
            <a>{{ extra }}</a>
          </slot>
        </div>
      </div>
      <div class="ix-card__body">
        <slot></slot>
      </div>
      <div v-if="isShowFooter" class="ix-card__footer">
        <slot name="footer"></slot>
      </div>
    </div>
  </div>
</template>
<script lang="ts">
import type { PropType } from 'vue'
import type { ComponentSize } from '@idux/components/core/types'
import type { CardConfig } from '@idux/components/core/config/types'

import { useGlobalConfig } from '@idux/components/core/config'
import { computed, defineComponent } from 'vue'
import { CardProps } from './types'

export default defineComponent({
  name: 'IxCard',
  props: {
    title: { type: String, default: undefined },
    extra: { type: String, default: undefined },
    hoverable: Boolean,
    borderless: Boolean,
    loading: Boolean,
    size: { type: String as PropType<ComponentSize>, default: undefined },
  },
  setup(props: CardProps, { slots }) {
    // init
    const cardConfig = useGlobalConfig('card')
    const classes = useClasses(props, cardConfig)
    const isShowLoading = computed(() => (props.loading !== undefined ? props.loading : cardConfig.loading))
    const isShowTitle = computed(() => typeof props.title === 'string' || !!slots.title)
    const isShowExtra = computed(() => typeof props.extra === 'string' || !!slots.extra)
    const isShowFooter = computed(() => !!slots.extra)
    return { classes, isShowTitle, isShowExtra, isShowFooter, isShowLoading }
  },
})

const useClasses = (props: CardProps, cardConfig: CardConfig) => {
  return computed(() => {
    const hoverable = props.hoverable !== undefined ? props.hoverable : cardConfig.hoverable
    const borderless = props.borderless !== undefined ? props.borderless : cardConfig.borderless
    const size = props.size !== undefined ? props.size : cardConfig.size
    return [
      size !== 'medium' ? `ix-card-${size}` : '',
      {
        'ix-card-hover': hoverable,
        'ix-card-border': borderless,
      },
    ]
  })
}
</script>
