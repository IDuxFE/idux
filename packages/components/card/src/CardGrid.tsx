/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, inject, normalizeClass } from 'vue'

import { useGlobalConfig } from '@idux/components/config'
import { IxCol } from '@idux/components/grid'

import { cardToken } from './token'
import { cardGridProps } from './types'

export default defineComponent({
  name: 'IxCardGrid',
  props: cardGridProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-card-grid`)
    const { hoverable } = inject(cardToken)!

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-hoverable`]: props.hoverable ?? hoverable.value,
      })
    })

    return () => <IxCol class={classes.value}>{slots.default?.()}</IxCol>
  },
})
