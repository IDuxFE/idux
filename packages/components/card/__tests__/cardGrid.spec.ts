import { flushPromises, mount } from '@vue/test-utils'

import Card from '../src/Card'
import CardGrid from '../src/CardGrid'

describe('CardGrid', () => {
  test('render work', async () => {
    const wrapper = mount({
      components: { Card, CardGrid },
      template: `
      <Card>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
      </Card>
      `,
    })
    await flushPromises()

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('hoverable work', async () => {
    const wrapper = mount({
      components: { Card, CardGrid },
      template: `
      <Card>
        <CardGrid hoverable>Content</CardGrid>
        <CardGrid hoverable>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid hoverable >Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
      </Card>
      `,
    })
    await flushPromises()

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('hoverable2 work', async () => {
    const wrapper = mount({
      components: { Card, CardGrid },
      template: `
      <Card hoverable>
        <CardGrid :hoverable="false">Content</CardGrid>
        <CardGrid :hoverable="false">Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
        <CardGrid>Content</CardGrid>
      </Card>
      `,
    })
    await flushPromises()

    expect(wrapper.html()).toMatchSnapshot()
  })
})
