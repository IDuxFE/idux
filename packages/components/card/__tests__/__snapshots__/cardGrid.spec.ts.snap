// Vitest Snapshot v1

exports[`CardGrid > hoverable work 1`] = `
"<div class=\\"ix-card ix-card-has-shadow ix-card-has-grid ix-card-sm\\">
  <!---->
  <!---->
  <div class=\\"ix-card-body\\">
    <div class=\\"ix-row\\">
      <div class=\\"ix-col ix-card-grid ix-card-grid-hoverable\\">Content</div>
      <div class=\\"ix-col ix-card-grid ix-card-grid-hoverable\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid ix-card-grid-hoverable\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
    </div>
  </div>
  <!---->
  <!---->
</div>"
`;

exports[`CardGrid > hoverable2 work 1`] = `
"<div class=\\"ix-card ix-card-has-shadow ix-card-has-grid ix-card-sm\\">
  <!---->
  <!---->
  <div class=\\"ix-card-body\\">
    <div class=\\"ix-row\\">
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid ix-card-grid-hoverable\\">Content</div>
      <div class=\\"ix-col ix-card-grid ix-card-grid-hoverable\\">Content</div>
      <div class=\\"ix-col ix-card-grid ix-card-grid-hoverable\\">Content</div>
      <div class=\\"ix-col ix-card-grid ix-card-grid-hoverable\\">Content</div>
      <div class=\\"ix-col ix-card-grid ix-card-grid-hoverable\\">Content</div>
    </div>
  </div>
  <!---->
  <!---->
</div>"
`;

exports[`CardGrid > render work 1`] = `
"<div class=\\"ix-card ix-card-has-shadow ix-card-has-grid ix-card-sm\\">
  <!---->
  <!---->
  <div class=\\"ix-card-body\\">
    <div class=\\"ix-row\\">
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
      <div class=\\"ix-col ix-card-grid\\">Content</div>
    </div>
  </div>
  <!---->
  <!---->
</div>"
`;
